### PR TITLE
fda: New command `heap-profile`.

### PR DESCRIPTION
**Pull request by blp**

### CODE REVIEW

**Comment by lalithsuresh** [comment]
If we cannot obtain one on aarch64/macs, we should return a friendly error message.

File: crates/fda/src/main.rs [hunk 1]
```diff
@@ -634,6 +636,62 @@ async fn pipeline(action: PipelineAction, client: Client) {
...
+                None => {
+                    let mut tempfile = tempfile().unwrap_or_else(|e| {
+                        eprintln!("ERROR: Failed to create temporary file: {e}");
+                        std::process::exit(1);
+                    });
+                    tempfile.write_all(&buffer).unwrap();
+
+                    let mut args = pprof.split_whitespace();
+                    let mut command =
+                        Command::new(args.next().expect("`pprof` command cannot be empty"));
```
Comment (gz) [inline crates/fda/src/main.rs:677]: Can we do a check if pprof is installed first and abort with an error giving guidance on how to install it? (as is already done in the command help)

**Review by gz — APPROVED** [review]

**Comment by blp** [comment]
> If we cannot obtain one on aarch64/macs, we should return a friendly error message.

I don't know how to do that. `heap_profile()` in the pipeline's server returns an appropriate error then the pipeline manager treats it as unexpected and the cli just bails without printing it. I guess the pipeline manager should pass it along but I don't know what the problem is.

**Comment by gz** [comment]
>I don't know how to do that. heap_profile() in the pipeline's server returns an appropriate error then the pipeline manager treats it as unexpected and the cli just bails without printing it. I guess the pipeline manager should pass it along but I don't know what the problem is.

This usually happens when the error code/response isn't properly decalred in the OpenAPI/utoipia annotations

File: crates/fda/src/main.rs [hunk 2]
```diff
@@ -1,7 +1,8 @@
 //! A CLI App for the Feldera REST API.
 
 use std::convert::Infallible;
-use std::io::Read;
+use std::io::{Read, Write};
```
[the remark this answers is not in the document]
Comment (blp) [reply crates/fda/src/main.rs:4]: What's the right way to use `tempfile` with tokio?

**Comment by blp** [comment]
> > I don't know how to do that. heap_profile() in the pipeline's server returns an appropriate error then the pipeline manager treats it as unexpected and the cli just bails without printing it. I guess the pipeline manager should pass it along but I don't know what the problem is.
> 
> This usually happens when the error code/response isn't properly decalred in the OpenAPI/utoipia annotations, you can check some other HTTP docs as a reference usually you need to list status code that you get and ErrorResponse as a type

I need help, I really don't know what I'm supposed to do for this.

**Comment by gz** [comment]
> I need help, I really don't know what I'm supposed to do for this.

no worries I can take a look

File: crates/fda/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gz) [reply crates/fda/src/main.rs:4]: Hm maybe we can just use this https://docs.rs/crate/async-tempfile/latest

**Comment by blp** [comment]
> > I need help, I really don't know what I'm supposed to do for this.
> 
> no worries I can take a look, fwiw we can create a separate issue and merge this now and fix it later

OK, I queued to merge and filed issue #2563 .

**Pull request merged**